# MetaMap NLP content view
This notebook intends to investigate the utility of MetaMap NLP content view (CVF=256) which claims to help NLP research. \
Here's a complete list of content views:
https://www.nlm.nih.gov/research/umls/knowledge_sources/metathesaurus/release/content_views.html

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
mrconso = pd.read_csv('/Users/qinyilong/Desktop/ScAi/MRCONSO.RRF', sep='|', header=None, dtype=object)

Meaning of columns: https://www.ncbi.nlm.nih.gov/books/NBK9685/table/ch03.T.concept_names_and_sources_file_mr/?report=objectonly

In [3]:
mrconso = mrconso.drop(18, axis=1) # last column is meaningless because the entry ends with '|'
mrconso.columns = ["CUI", "LAT", "TS", "LUI", "STT", "SUI", "ISPREF", "AUI", "SAUI", "SCUI", "SDUI", "SAB", "TTY", "CODE", "STR", "SRL", "SUPPRESS", "CVF"]

In [4]:
# Extract entries within MetaMap view
mrconso_metamap_on = mrconso[mrconso.CVF == "256"]

In [5]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_on.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
4144944   C0449546  ENG  P   L0800556  PF   S0849933      N   A0908656   
12005268  C3432640  ENG  P  L10685815  PF  S13205143      Y  A20567502   
3124965   C0329058  ENG  P   L0516177  PF   S0744100      N   A0804924   
9132674   C1936187  ENG  P   L6658440  PF   S7688728      Y  A12919250   
8980168   C1850055  ENG  P   L6468075  PF  S11794726      Y  A18470743   
2880931   C0301304  ENG  P   L0598641  VO  S13956832      N  A23075124   
12743769  C3829001  ENG  S  L11665161  PF  S14461537      Y  A23939462   
8663880   C1666686  ENG  S   L6395571  PF   S7327381      N  A23088241   
10573547  C2635433  ENG  P   L8651374  PF  S10723554      Y  A16598591   
2228427   C0222167  ENG  S   L2848512  PF   S3546145      Y   A3708206   
3948436   C0428481  ENG  P   L3506084  PF   S4044893      Y   A4373913   
11529763  C3101350  ENG  P   L9737897  PF  S12247009      Y  A19158863   
3195296   C0334526  ENG  P   L0536398  PF   S0739765      N   A0800698   
7985692   C1423635  ENG  P   L5139975  PF   S5874570      N  A20703337   
13518020  C4225443  ENG  P  L13343135  PF  S16300513      Y  A26957633   
5840391   C0833923  ENG  S   L1671056  PF   S1881916      Y  A15439256   
296265    C0010823  ENG  S   L0021161  VO   S0051980      Y   A0072980   
8008583   C1426611  ENG  S   L7580357  PF   S8775920      Y  A20739865   
8633374   C1648239  ENG  P   L5804621  PF   S6610945      Y   A9290951   
11235586  C2925995  ENG  P   L9450416  PF  S11744363      Y  A18278811   
6660100   C1011727  ENG  P   L2002714  PF   S2302204      Y   A2151564   
11590003  C3152187  ENG  P   L9947265  VO  S12430869      Y  A19292575   
13969845  C4470113  ENG  P  L14001722  PF  S16977618      Y  A28196451   
3227529   C0337974  ENG  S   L0521404  PF   S0661931      Y   A3030976   
953604    C0033126  ENG  P   L0033126  PF   S0076313      N  A22911989   
6649886   C1008029  ENG  P   L2005061  PF   S2304655      Y   A2154047   
2107537   C0201983  ENG  S   L2990542  VO   S3251184      N  A22992010   
4653304   C0529155  ENG  S   L0621801  PF   S1019600      Y   A1076879   
6678766   C1018605  ENG  P   L2005354  PF   S2304918      Y   A2154314   
3209117   C0335915  ENG  P   L0537831  PF   S0757693      Y   A3083830   
341788    C0012624  ENG  S   L0038014  VO   S0087644      Y   A0118787   
14039103  C4506954  ENG  P  L14166959  PF  S17237618      Y  A28386855   
5053917   C0591391  ENG  P   L0816564  PF  S11879448      Y  A18603722   
2235206   C0222719  ENG  S   L0308601  PF   S0380781      N   A2654698   
10740003  C2718002  ENG  P   L8909855  VO  S11095352      Y  A17004088   
2268220   C0225168  ENG  P   L2930093  PF   S3560700      N   A3723654   
1879154   C0175276  ENG  S   L1456547  VO   S1742436      Y  A22649496   
6719183   C1035905  ENG  S   L2034849  PF   S2343680      Y   A2193610   
2353483   C0230629  ENG  P   L0318950  PF   S0411744      N  A27868221   
10899279  C2773591  ENG  P   L9260133  PF  S11435185      Y  A17533495   
4693833   C0546255  ENG  P   L1013240  VO  S15900342      Y  A25772815   
13043670  C3951656  ENG  P  L11926406  PF  S14904660      Y  A24501309   
983216    C0034155  ENG  S   L0039207  PF   S0003788      Y   A0009877   
8266533   C1524442  ENG  S   L5392473  PF  S17148582      Y  A28280014   
2204963   C0220650  ENG  S   L0266707  VO   S8417433      Y  A13816694   
3266051   C0341184  ENG  P   L0552681  PF   S0624104      N  A25707611   
9106565   C1918144  ENG  S  L12018636  PF  S14863197      Y  A24494893   
13351942  C4090682  ENG  P  L13227748  PF  S16269439      Y  A26729151   
5199602   C0679403  ENG  P   L1198872  VO   S1437379      N   A1397225   
7071353   C1170062  ENG  P   L2490440  PF   S2941026      N   A7826354   
10553957  C2622830  ENG  P   L8687661  PF  S10748786      Y  A16524965   
9833679   C2193245  ENG  P   L8519014  PF  S10606158      N  A16357281   
12842847  C3865224  ENG  P  L11785366  PF  S14644999      N  A24218607   
13536369  C4236613 

In [6]:
# Frequencies of entries in different languages
mrconso_metamap_on["LAT"].value_counts()

ENG    4802688
Name: LAT, dtype: int64

In [7]:
sources_metamap_on = mrconso_metamap_on["SAB"].value_counts()

In [8]:
# Extract entries without MetaMap view
mrconso_metamap_off = mrconso[mrconso.CVF != "256"]

In [9]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_off.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
1939584   C0186085  ENG  S   L3108374  PF   S3595776      Y   A3760601   
11064088  C2858118  ENG  P   L9370743  PF  S11651081      N  A17857832   
5398577   C0717309  ENG  S   L9104624  PF  S11358553      Y  A17399200   
13112765  C4003259  ENG  S  L11905022  PF  S14891419      Y  A24499805   
7090285   C1177892  ENG  S  L11394887  PF  S17865454      Y  A29254884   
8159633   C1483328  ENG  S   L8691523  PF  S10780139      Y  A16510597   
12834106  C3861884  ENG  P  L11765940  PF  S15033924      Y  A24767918   
6152596   C0881262  ITA  P  L12890220  PF  S15339259      Y  A25101439   
3256160   C0340424  SPA  P   L4452293  PF   S5135763      Y   A5908758   
10943983  C2805902  ENG  S   L9266823  PF  S11488120      Y  A17510432   
9335754   C2013187  ENG  P   L7242237  PF   S8460764      Y  A14150034   
9454907   C2055707  ENG  P   L7491514  PF   S8653787      Y  A14024721   
13135022  C4019435  SPA  S  L12989891  PF  S15932505      Y  A29287959   
9877058   C2209913  ENG  P   L9016782  PF  S11214436      Y  A17253698   
14354185  C4677557  ENG  S  L14693376  PF  S17757192      Y  A28998045   
632845    C0021970  NOR  S  L11441860  PF  S14213383      Y  A27671372   
14173642  C4548296  ENG  S  L14542036  PF  S17642704      Y  A28921544   
229763    C0008372  SPA  P   L4003739  PF   S4687223      Y   A5450092   
9942629   C2234373  ENG  S   L7408943  PF  S10974104      Y  A16801182   
8094472   C1449832  JPN  S   L8010841  PF  S10067149      Y  A15682374   
12132086  C3506447  ENG  P  L10901216  PF  S13567597      Y  A21019898   
2504543   C0242917  GER  S  L11672227  PF  S14485855      N  A25918314   
10324238  C2444175  ENG  S   L8214408  PF  S10352100      Y  A16209196   
1260116   C0045869  FRE  S   L5703434  PF   S6536472      Y  A27695592   
11557690  C3126578  ENG  P   L9773648  PF  S12335244      Y  A19188075   
5230332   C0685874  JPN  S   L6269839  PF   S7211353      N  A26346306   
6746926   C1049629  ENG  S   L5758175  PF   S6585944      Y  A27947989   
2397446   C0233956  ENG  S   L7842949  PF   S9092871      Y  A15266434   
3751913   C0408670  ENG  S   L2987707  PF   S3247245      Y  A22892286   
8116063   C1456606  FIN  P   L6940784  PF   S8079998      Y  A13431704   
5359045   C0707741  ENG  S   L2434454  PF   S2888823      Y  A10444661   
1934815   C0185569  ENG  S   L0223932  VO  S10202676      Y  A15995386   
11058238  C2856100  ENG  S  L14942326  PF  S18191597      Y  A29634678   
5908889   C0853474  DUT  P   L4981160  PF   S5668664      N  A25828406   
9940815   C2233706  ENG  S   L7892056  PF   S9141329      Y  A15310535   
7458167   C1289483  ENG  S   L2901821  PF   S3283143      Y   A3421295   
12966900  C3898424  ENG  S  L11878219  VO  S16093117      Y  A26712270   
11008501  C2837275  ENG  P   L9393901  PF  S11675778      Y  A29699497   
3393687   C0355700  ENG  S   L2763605  PF   S3479387      Y   A5021640   
1303276   C0055408  ENG  S   L0101142  PF   S0120041      Y   A0151288   
6284685   C0936005  SWE  P  L13743017  PF  S16754528      Y  A27692696   
12955176  C3894683  RUS  S  L13068628  PF  S16010022      Y  A26533756   
10830680  C2740073  ENG  S   L5858552  PF  S14504810      Y  A24026590   
10314727  C2439942  ENG  S  L10224863  PF  S12800163      Y  A19899034   
1903865   C0179820  SPA  P   L3979652  PF   S4663122      Y   A5425233   
907353    C0031502  ITA  P   L2139636  PF   S2477858      Y   A7551035   
8329224   C1534836  ENG  P   L7771470  PF   S9015007      Y  A15735963   
691945    C0023761  RUS  S   L0895318  PF   S1099135      Y  A13484093   
6580626   C0981626  ENG  S   L0472426  PF   S0694641      Y   A0751009   
4375528   C0480374  ENG  S   L5844990  PF   S6700364      Y  A23040371   
11079722  C2863672  ENG  P   L9350060  PF  S11650767      N  A17820437   
11793386  C3260843  ENG  S  L10958438  PF  S13973372      Y  A22758242   
7563132   C1305489  SPA  S   L3901741  PF   S4585209      Y   A5344726   
10597171  C2653433 

In [10]:
# Frequencies of entries in different languages
mrconso_metamap_off["LAT"].value_counts()

ENG    5508879
SPA    1470221
FRE     424471
JPN     322983
POR     300417
DUT     286526
ITA     236976
GER     232823
RUS     180945
CZE     180525
HUN     104811
CHI      77622
NOR      61724
TUR      51114
POL      50108
KOR      38660
EST      31204
SWE      29737
FIN      25486
SCR       9981
GRE       2241
LAV       1405
DAN        723
BAQ        695
HEB        485
Name: LAT, dtype: int64

In [11]:
sources_metamap_off = mrconso_metamap_off["SAB"].value_counts()

In [12]:
# Let's what english words are excluded from MetaMap view
mrconso_metamap_off_eng = mrconso_metamap_off[mrconso_metamap_off.LAT == "ENG"]

In [13]:
# Randomly sample 100 entries and display
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(mrconso_metamap_off_eng.sample(100))

CUI  LAT TS        LUI STT        SUI ISPREF        AUI  \
5407999   C0720580  ENG  S   L6685369  PF   S7783691      Y  A13271797   
9678933   C2137556  ENG  P   L9048167  PF  S11188184      Y  A17220512   
12280226  C3573596  ENG  S  L11161815  PF  S13764253      Y  A21353725   
10221166  C2384273  ENG  P   L8270604  PF  S10501672      Y  A16330708   
14293578  C4622182  ENG  P  L14680706  PF  S17830071      Y  A29052927   
12447436  C3667081  ENG  S  L11688438  PF  S14495436      Y  A24035421   
3311653   C0344928  ENG  S   L2988479  PF   S3473499      Y   A3625058   
5056591   C0592445  ENG  S   L1046707  PF   S1260990      Y   A1219947   
13338419  C4087190  ENG  P  L13135913  PF  S16096190      Y  A26715362   
11444205  C3022403  ENG  P   L9867468  PF  S12285066      Y  A19173416   
4540249   C0498126  ENG  P   L0808498  PF  S13908689      Y  A21437425   
14006565  C4492790  ENG  P  L14133072  PF  S17187615      Y  A28320771   
9723901   C2154301  ENG  P   L7560864  PF   S8598555      Y  A13832553   
11336815  C2962925  ENG  S  L10461902  PF  S13055849      Y  A20283593   
10420810  C2518607  ENG  P   L8436719  PF  S10341591      Y  A19887355   
7492901   C1294562  ENG  S  L14915752  PF  S18112711      Y  A29490920   
4701876   C0548803  ENG  P   L1029985  PF   S1242575      Y   A1206221   
13374106  C4100743  ENG  S  L13201544  PF  S16246289      Y  A26764778   
11647630  C3173734  ENG  P  L10091992  PF  S12602607      N  A19532683   
13823678  C4352365  ENG  P  L13991877  PF  S17092114      Y  A28127753   
1849327   C0162596  ENG  S   L0508025  VO   S1649688      Y   A1611881   
5406766   C0719997  ENG  P   L1337168  VO   S8145879      Y  A13734871   
10720148  C2712730  ENG  P   L8907532  VO  S12053133      Y  A18924083   
257090    C0009400  ENG  P   L0009400  VO   S4126021      Y   A4458643   
12153124  C3515614  ENG  P  L10909391  PF  S13595554      N  A23621862   
4925636   C0575015  ENG  S   L3064842  PF   S3412977      Y   A3559630   
1952596   C0187410  ENG  S  L14183413  PF  S17282809      N  A28464250   
5582129   C0780648  ENG  P   L5331040  VO  S17896234      Y  A29278775   
14420421  C4711110  ENG  P  L14924664  PF  S18142908      Y  A29579476   
14396296  C4700350  ENG  S  L14817486  PF  S18023806      Y  A29410589   
7376492   C1278130  ENG  S   L7775767  VO  S11063748      Y  A16958203   
11211222  C2919791  ENG  S   L9438279  PF  S11723693      Y  A24253049   
12908606  C3886832  ENG  P  L11856119  PF  S14728565      Y  A24325619   
7524217   C1299454  ENG  S   L2813550  PF   S3469553      Y   A3620626   
13123303  C4012270  ENG  P  L12047465  PF  S14939849      Y  A24620353   
9416388   C2042355  ENG  S   L9039932  PF  S11225569      Y  A17142218   
6273206   C0931659  ENG  S   L1804782  PF   S2118462      Y  A27052264   
14175533  C4549559  ENG  S   L6434459  PF  S17848000      Y  A29144676   
5372434   C0710299  ENG  S   L9983272  VO  S17117972      Y  A28201239   
11257986  C2933831  ENG  P   L9481227  PF  S11782599      Y  A18468792   
11232929  C2924702  ENG  P  L11241965  PF  S13963756      Y  A22743188   
11196899  C2910268  ENG  S  L10362819  PF  S12921328      Y  A20154062   
9039683   C1877226  ENG  P   L6556435  VO   S7647245      Y  A12677935   
3420090   C0361331  ENG  P   L0483308  PF   S0744392      Y   A3883660   
9853031   C2200750  ENG  S   L8800371  PF  S10968304      Y  A16889692   
8433468   C1556362  ENG  S  L14743177  PF  S17838206      Y  A29163223   
9918503   C2224925  ENG  S   L8176845  PF  S10220041      Y  A15924920   
6874747   C1113811  ENG  S   L8549034  PF  S10618799      Y  A18147343   
10379163  C2474757  ENG  P   L8465722  PF  S10575968      Y  A19882710   
8797362   C1716337  ENG  S   L8571518  PF  S10677041      Y  A18879778   
9834848   C2193702  ENG  S   L7372571  PF   S8276105      Y  A13953686   
10879566  C2761143  ENG  S   L9216926  PF  S11521541      Y  A28093229   
2750482   C0274211  ENG  S   L6983449  PF   S8186314      Y  A14115316   
12071141  C3483230 

In [14]:
sources = pd.DataFrame(dict(included = sources_metamap_on, excluded = sources_metamap_off))

In [15]:
# Frequencies of entries from different source vocabularies
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(sources)

included  excluded
AIR                  337.0       347
ALT                  200.0      8865
AOD                17933.0      2746
AOT                  444.0        27
ATC                 5506.0      1310
BI                   951.0       300
CCC                   62.0       348
CCPSS               1764.0     14070
CCS                  936.0       681
CCS_10               444.0       436
CDT                   13.0       785
CHV               125619.0     20533
COSTAR              3339.0       119
CPM                  446.0      2649
CPT                 2557.0     68333
CPTSP                  NaN      2684
CSP                21674.0      1093
CST                 4263.0      2179
CVX                  234.0       389
DDB                   29.0       227
DMDICD10               NaN     11999
DMDUMD                 NaN      3339
DRUGBANK           15726.0      7454
DSM-5                236.0      1101
DXP                 9109.0      1003
FMA               156994.0     14003
GO                118786.0     60134
GS                   570.0     61066
HCDT                  32.0      1378
HCPCS                878.0     12300
HCPT                 405.0     20742
HGNC              154424.0     19307
HL7V2.5             3114.0      1905
HL7V3.0             9560.0      2442
HPO                27221.0      3552
ICD10               5571.0      7929
ICD10AE              459.0       648
ICD10AM             7517.0     18364
ICD10AMAE            749.0      1656
ICD10CM            38516.0    140534
ICD10DUT               NaN     11359
ICD10PCS           12992.0    336914
ICD9CM             14635.0     26178
ICF                  730.0       791
ICF-CY               825.0       946
ICNP                 178.0      1863
ICPC                 640.0       413
ICPC2EDUT              NaN       685
ICPC2EENG            542.0       837
ICPC2ICD10DUT          NaN     76614
ICPC2ICD10ENG       2323.0     79451
ICPC2P              4765.0     12124
ICPCBAQ                NaN       695
ICPCDAN                NaN       723
ICPCDUT                NaN       723
ICPCFIN                NaN       722
ICPCFRE                NaN       723
ICPCGER                NaN       723
ICPCHEB                NaN       485
ICPCHUN                NaN       718
ICPCITA                NaN       723
ICPCNOR                NaN       722
ICPCPOR                NaN       723
ICPCSPA                NaN       723
ICPCSWE                NaN       723
JABL                 530.0      2730
KCD5                   NaN     10997
LCH                 6407.0       244
LCH_NW             12533.0       795
LNC               203805.0    275488
LNC-DE-AT              NaN      5814
LNC-DE-CH              NaN      4941
LNC-DE-DE              NaN     11059
LNC-EL-GR              NaN      2241
LNC-ES-AR              NaN     76586
LNC-ES-CH              NaN      4940
LNC-ES-ES              NaN     54206
LNC-ET-EE              NaN     31204
LNC-FR-BE              NaN     45905
LNC-FR-CA              NaN     45349
LNC-FR-CH              NaN      4940
LNC-FR-FR              NaN     96894
LNC-IT-CH              NaN      4939
LNC-IT-IT              NaN     66396
LNC-KO-KR              NaN     26893
LNC-NL-NL              NaN     52516
LNC-PT-BR              NaN    110887
LNC-RU-RU              NaN     57499
LNC-TR-TR              NaN     51114
LNC-ZH-CN              NaN     77622
MCM                   39.0         4
MDR                29717.0     75619
MDRCZE                 NaN    104092
MDRDUT                 NaN    104092
MDRFRE                 NaN    104092
MDRGER                 NaN    104092
MDRHUN                 NaN    104093
MDRITA                 NaN    104094
MDRJPN                 NaN    198936
MDRPOR                 NaN    104094
MDRSPA                 NaN    104094
MED-RT              6630.0      1213
MEDCIN             21483.0    879813
MEDLINEPLUS         2986.0       393
MMSL               10877.0     71266
MMX                 1513.0     71119
MSH               609456.0    262484
MSHCZE                

In [20]:
# Some shapes
print("Entire MRCONSO: {}".format(mrconso.shape))
print("MetaMap view: {}".format(mrconso_metamap_on.shape))
print("MetaMap view excluded: {}".format(mrconso_metamap_off.shape))
print("English entries excluded by MetaMap view: {}".format(mrconso_metamap_off_eng.shape))

Entire MRCONSO: (14433450, 18)
MetaMap view: (4802688, 18)
MetaMap view excluded: (9630762, 18)
English entries excluded by MetaMap view: (5508879, 18)


### Observations:
Metamap NLP content view extracts 4802688 english concepts and discards 5508879 english concepts and all the non-english concepts. \
Source Vocabularies like LONIC (LNC-), which are suggested by reference manual to be detrimental to NLP, are completely excluded. \

### TODO:
What kinds of English concepts were excluded?